From: Gerald Schaefer <geraldsc@de.ibm.com>
Subject: qeth (new function): tolerate OLM-limitation
References: bnc#700080,LTC#69475,FATE#311892
Patch-mainline: Yes

Description: z/OS may activate Optimized Latency Mode (OLM) for a
             connection through an OSA Express3 adapter, which
             reduces the number of allowed concurrent connections,
             if adapter is used in shared mode.
             Create a meaningful message, if activation of an
             OSA-connection fails due to an active OLM-connection
             on the shared OSA-adapter.

Acked-by: John Jolly <jjolly@suse.de>

---
 drivers/s390/net/qeth_core_main.c |    7 +++++++
 1 file changed, 7 insertions(+)

--- a/drivers/s390/net/qeth_core_main.c
+++ b/drivers/s390/net/qeth_core_main.c
@@ -2057,6 +2057,13 @@ static int qeth_ulp_setup_cb(struct qeth
 	memcpy(&card->token.ulp_connection_r,
 	       QETH_ULP_SETUP_RESP_CONNECTION_TOKEN(iob->data),
 	       QETH_MPC_TOKEN_LENGTH);
+	if (!strncmp("00S", QETH_ULP_SETUP_RESP_CONNECTION_TOKEN(iob->data),
+		     3)) {
+		QETH_DBF_TEXT(SETUP, 2, "olmlimit");
+		dev_err(&card->gdev->dev, "A connection could not be "
+			"established because of an OLM limit\n");
+		iob->rc = -EMLINK;
+	}
 	QETH_DBF_TEXT_(SETUP, 2, "  rc%d", iob->rc);
 	return 0;
 }
